readme.md
This is an early version of my first platformer - Crabby Patti vs The World!
It is a clone from an old Mario game. Hit platforms from below to flip
enemies and then go kick them before they recover! Save your crab friends
from the evil scaly hands of corporate reptilians!

CREDITS

The main character (Crabby Patti) was designed by @van___van (instagram)
and adapted to pixel art by me!

-
Music:
 - Title score by HitCtrl (https://opengameart.org/content/hitctrl-happy-pixie-town)
 - Level 1 - https://opengameart.org/content/clouds-castle-8-bit-videogame-loop
 - Level 2 - https://opengameart.org/content/apocalyptic-fire-town
 - Level 3 - https://opengameart.org/content/law-in-the-city

-
As for me I'm a boring slave to the system by day, storyteller and (newb) indie game dev by
night and you can find me around the webz (instagram,github,twitter,itch.io) @amirani.otm

For inquiries or comments hit me at hector.m0202@gmail.com


changes.txt
##### VERSION 1.0.4 ######

Fully playable!

OS:
 - changing folder system

mechs:
  - 

beautify:
  -

to fix:
  - sound playing


tilemap.py
import pygame as pg
from settings import *
vec = pg.math.Vector2

#collides hit rectangles of sprites instead of sprite images
#why is it in tilemap?
def collide_hit_rect(one,two):
    return one.hit_rect.colliderect(two.rect)

#game map created by parsing through map file (txt)
class Map:
    def __init__(self,filename):
        #filling a list with -letters?- that is passed on to the game to draw the map
        self.data = []
        with open(filename,'rt') as f:
            for line in f:
                self.data.append(line.strip())
        #setting map and tile size
        self.tilewidth = len(self.data[0])
        self.tileheight = len(self.data)
        self.width = self.tilewidth * tilesize
        self.height = self.tileheight* tilesize

#camera keeps player always centered on screen
class Camera:
    def __init__(self, mywidth, myheight):
        self.camera = pg.Rect(0, 0, mywidth, myheight)
        self.width = mywidth
        self.height = myheight

    #shifts every sprite around player movement
    def apply(self, entity):
        return entity.rect.move(self.camera.topleft)

    def update(self, target):
        #follows player position around
        x = -target.rect.centerx + int(width / 2)
        y = -target.rect.centery + int(height / 2)

        # limits scrolling to map size
        x = min(0, x)  # left
        y = min(0, y)  # top
        x = max(-(self.width - width), x)  # right
        y = max(-(self.height - height), y)  # bottom
        self.camera = pg.Rect(x, y, self.width, self.height)


settings.py
import os
import math
import pygame as pg

# defining colors
white = (255,255,255)
black = (0,0,0)
darkgrey = (40,40,40)
lightgrey = (100,100,100)
green = (0,255,0)
red = (255,0,0)
blue = (0,0,255)
yellow = (255,255,0)

# Game settings
width = 1024
height = 704
fps = 60
title = "Crabby Patti vs The World"
bgcolor = black
#test enemies
'''enemies = [[{'flamey':1,'pgrunt':1}],[{'pgrunt':1, 'psuper':1,'rept':1}],[{'psuper':1,'rept': 1, 'baby':1}]]'''
#final enemies
enemies = [[{'pgrunt':5},{'pgrunt':4, 'psuper':2},{'pgrunt':2, 'psuper':4}],
            [{'pgrunt':4, 'psuper':4,}, {'pgrunt':4, 'psuper':4, 'flamey':1}, {'pgrunt':2, 'psuper':3,'rept':2,'flamey':2}],
            [{'psuper':2,'rept': 4, 'baby':1,'flamey':1}, {'psuper':2,'rept':4,'baby':2,'flamey':1}, {'rept':4, 'baby':4,'flamey':2}]]
enemy_types = ['pgrunt','psuper','flamey','rept','baby']
transition_time = 3000
level_time = 90
pb_pows = 3
walk_delay = 25
total_levels = 3

# Map settings
tilesize = 32
#gridwidth = width/tilesize
#gridheight = height/tilesize
#plat settings
plat_burn_delay = 50

# Player settings
player_acc = 6
player_fric = -1
player_grav = 1.8
player_jump = -35
player_image = ''
player_lives = 3
player_respawn_time = 3000
player_jump_delay = 50
player_input_delay = 4000

# Mob settings
mob_image = ''
mob_hit_rect = pg.Rect(0, 0, 32, 32)
mob_speed = 1.8
mob_spawn_vel = 1
mob_spawn_time = 4000
mob_flip_time = 4500
mob_mad_time = 3000
mob_jump = -2
mob_jump_delay = 1500
mob_jump_time = 400
mob_flame_time = 1500
mob_tremble_delay = 400

# Folders
game_fld = os.path.dirname(__file__)
media_fld = os.path.join(game_fld, 'media')
img_fld = os.path.join(media_fld, 'sprites')
blocks_fld = os.path.join(media_fld, 'blocks')
font_fld = os.path.join(media_fld, 'fonts')
sound_fld = os.path.join(media_fld, 'sfx')


main.py
import pygame as pg
import random
import sys
import six
#only for packing
#import packaging
from settings import *
from sprites import *
from tilemap import *

# Classes
class Game:
    #main class of the game, contains and initializes fundamental objects (clock, screen, etc) and controls load, main loop and quitting of the game
    def __init__(self):
        pg.init()
        #pg.mixer.pre_init()
        pg.mixer.init()
        self.screen = pg.display.set_mode((width, height))
        self.clock = pg.time.Clock()
        self.paused = False
        self.cdown = level_time
        self.level = 1
        self.playing = False
        self.lives_spent = 0
        self.start_tick = 0
        self.player_score = 0
        self.started = False
        self.load_data()

    #initial data loading. Folder mapping can be done here but is done instead in the settings file
    #game_fld = os.path.dirname(__file__)
    #img_fld = os.path.join(game_fld, 'sprites')
    #player_fld = os.path.join(img_fld, 'char')

    def load_data(self):
        self.spritesheet = Spritesheet(os.path.join(img_fld,'spritesheet.png'),self)
        with open(os.path.join(game_fld,'highscore.txt'),'w') as f:
            try:
                self.highscore = int(f.read())
            except:
                self.highscore = 0

    #initializes new objects: sprite groups and sprites
    def new(self,level):
        self.map = Map(os.path.join(game_fld,'map32.txt'))
        self.all_sprites = pg.sprite.LayeredUpdates()
        self.players = pg.sprite.Group()
        self.utilities = pg.sprite.Group()
        self.lethals = pg.sprite.Group()
        self.counters = pg.sprite.Group()
        self.platforms = pg.sprite.Group()
        self.mobs = pg.sprite.Group()
        for row, tiles in enumerate(self.map.data):
            for col, tile in enumerate(tiles):
                if tile in ['0','1','2','3','4','5','6','7','8','9']:
                    Platform(self, col, row, tile)
                if tile == 'B':
                    if level == 1:
                        if row < 16:
                            self.pb1 = PowBlock(self,col,row)
                        else:
                            self.pb2 = PowBlock(self,col,row)
                if tile == 'P':
                    self.player = Player(self, col, row)
                if tile == 'S':
                    if col < 16:
                        self.sp1 = SpawnPoint(self, col, row, True)
                    else:
                        self.sp2 = SpawnPoint(self, col, row, False)
                if tile == 'C':
                    self.counter = Counter(self, col, row)
        self.camera = Camera(self.map.width, self.map.height)
        self.spawnevent = pg.USEREVENT + 1
        pg.time.set_timer(self.spawnevent, mob_spawn_time)
        if self.level > 1:
            if self.pb1.pows > 0:
                self.all_sprites.add(self.pb1)
            if self.pb2.pows > 0:
                self.all_sprites.add(self.pb2)
        if self.level == 1:
            self.bg = self.spritesheet.get_image('beach_bg')
        elif self.level == 2:
            self.bg = self.spritesheet.get_image('volcano_bg')
        elif self.level == 3:
            self.bg = self.spritesheet.get_image('city_bg')

    #standard finishing method - as seen on the internet
    def quit(self):
        pg.quit()
        sys.exit()

    #executes every single sprite update routine and controls where (to whom) camera is pointed
    def update(self):
        self.all_sprites.update()
        self.camera.update(self.player)
        if self.started:
            self.dt = self.clock.tick(fps) / 1000.0
            self.cdown -= self.dt
            if self.cdown <= 0:
                self.playing = False

    #guide grid can be drawn through this method
    '''def draw_grid(self):
        for x in range(0, width, tilesize):
            pg.draw.line(self.screen, lightgrey, (x,0),(x,height))
        for y in range(0, height, tilesize):
            pg.draw.line(self.screen, lightgrey, (0,y), (width,y))'''

    def draw_text(self, text, size, color, x, y):
        self.font = pg.font.Font(os.path.join(font_fld, 'Ubuntu-B.ttf'), size)
        text_surface = self.font.render(text, True, color)
        text_rect = text_surface.get_rect()
        text_rect.midtop = (x,y)
        self.screen.blit(text_surface,text_rect)

    #sets graphical elements: display title (here - fps), sets screen background, and draws all sprites on the screen
    def draw(self):
        pg.display.set_caption('{:.2f}'.format(self.clock.get_fps()))
        self.screen.fill(bgcolor)
        self.screen.blit(pg.transform.scale(self.bg,(self.bg.get_width()*8,self.bg.get_height()*8)),(0,0))
        #self.cdown -= (pg.time.get_ticks()/1000)
        for sprite in self.all_sprites:
            self.screen.blit(sprite.image,self.camera.apply(sprite))
        if self.playing:
            case = self.spritesheet.get_image('case').convert_alpha()
            life = self.spritesheet.get_image('patti_life').convert_alpha()
            pts = self.spritesheet.get_image('pts').convert_alpha()
            spoint = self.spritesheet.get_image('spawn_point_1').convert_alpha()
            self.screen.blit(pg.transform.scale(case,(int(case.get_width()*2),int(case.get_height()))),((width/2)-30,30))
            self.draw_text(str('{:.1f}'.format(self.cdown)), 22, white, width/2,32)
            self.screen.blit(pg.transform.scale(case,(int(case.get_width()*2.7),int(case.get_height()*2.5))),(width-int(case.get_width()*2.7),30))
            self.screen.blit(pg.transform.scale(pts,(int(case.get_width()),int(case.get_height()*0.5))),(width-72,43))
            self.screen.blit(pg.transform.scale(life,(int(case.get_width()*0.9),int(case.get_height()*0.8))),(width-72,73))
            self.draw_text(str(self.player_score), 20, white, width-23, 40)
            self.draw_text(str(self.player.lives), 20, white, width-23, 74)
        #self.draw_grid()
        pg.display.flip()

    #overall process input control - eg keys that interrupt the game directly
    def events(self):
        for event in pg.event.get():
            if event.type == pg.QUIT:
                self.quit()
            if event.type == pg.KEYDOWN:
                if event.key == pg.K_ESCAPE:
                    self.quit()
                if (event.key == pg.K_SPACE or event.key == pg.K_z) and not (self.player.isjump or self.player.midair):
                    jump(self.player)
            if event.type == self.spawnevent:
                for sp in self.utilities:
                    if sp.active:
                        self.counter.send_mob(sp)
                        sp.active = False
                    else:
                        sp.active = True

    def wait_for_key(self):

        waiting = True
        while waiting:
            self.clock.tick(fps)
            for event in pg.event.get():
                if event.type == pg.QUIT:
                    waiting = False
                    self.quit()
                if event.type == pg.KEYDOWN:
                    if event.key == pg.K_ESCAPE:
                        self.quit()
                    else:
                        waiting = False
                        self.playing = True

    #title screen
    def show_start_screen(self):
        pg.mixer.music.stop()
        pg.mixer.music.load(os.path.join(sound_fld,'intro.ogg'))
        pg.mixer.music.play(-1)
        self.screen.fill(bgcolor)
        #self.draw_text(title, 48, white, width / 2, height / 4)
        banner = self.spritesheet.get_image('main_title')
        self.screen.blit(banner,((width / 3) - 10, height / 4))
        self.draw_text("Arrows/WASD to move, Space/Z to jump", 22, white, width / 2, height* 3 / 5)
        self.draw_text("Press a key to play", 22, white, width / 2, height * 3 / 4)
        self.draw_text('High Score: ' + str(self.highscore), 22, white, width/2, 15)
        pg.display.flip()
        self.wait_for_key()

    def show_go_screen(self, type):
        if type == 'over':
            self.screen.fill(bgcolor)
            if self.player.lives == 0 or self.cdown <= 0:
                #self.draw_text("GAME OVER", 48, white, width / 2, height / 4)
                cap = self.spritesheet.get_image('game_over')
                self.screen.blit(cap, (width / 3, height / 4))
            else:
                #self.draw_text("YOU WON!!", 48, white, width / 2, height / 4)
                cap = self.spritesheet.get_image('you_won')
                self.screen.blit(cap, (width / 3, height / 4))
            self.draw_text("Press a key to play again", 22, white, (width / 2)+30, height * 3 / 4)
            if self.player_score > self.highscore:
                self.highscore = self.player_score
                with open(os.path.join(game_fld,'highscore.txt'),'w') as f:
                    f.write(str(self.player_score))
                self.draw_text('New High Score! ' + str(self.highscore), 22, white, width/2, height/2)
            else:
                self.draw_text('Score: ' + str(self.player_score), 22, white, width/2, height/2 - 40)
                self.draw_text('High Score: ' + str(self.highscore),22,white,width/2,height/2)
            self.lives_spent = 0
            self.player_score = 0
            self.cdown = level_time
            self.level = 1
            pg.display.flip()
            self.wait_for_key()
        if type == 'level':
            pg.mixer.music.stop()
            if self.level == 1:
                pg.mixer.music.load(os.path.join(sound_fld,'level_1.ogg'))
                pg.mixer.music.set_volume(0.2)
            if self.level == 2:
                pg.mixer.music.load(os.path.join(sound_fld,'level_2.ogg'))
                pg.mixer.music.set_volume(0.4)
            if self.level == 3:
                pg.mixer.music.load(os.path.join(sound_fld,'level_3.ogg'))
                pg.mixer.music.set_volume(0.2)
            self.screen.fill(bgcolor)
            self.draw_text('Level ' + str(self.level), 38, white, width/2,height/2)
            pg.display.flip()
            pg.time.wait(3000)
            self.new(self.level)
            self.draw()
            self.counter.next_stage()
            pg.mixer.music.play(-1)

    #game main loop
    def run(self):
        self.show_start_screen()
        self.show_go_screen('level')
        while self.playing:
            self.update()
            self.events()
            self.draw()
        self.show_go_screen('over')

# Creating game object
g = Game()
while True:
    g.run()
